sticker-utils/src/bin/sticker-pretrain.rs

```diff
@@ -0,0 +1,346 @@
+                &app,
+                trainer,
+                RelativePositionEncoder,
+                train_file,
+                validation_file,
+            )
+        }
+    }
+    .or_exit("Error while training model", 1);
+}
+
+fn create_trainer(config: &Config, app: &PretrainApp) -> Fallible<TaggerTrainer> {
+    let graph_read = BufReader::new(File::open(&config.model.graph)?);
+    let graph = TaggerGraph::load_graph(graph_read, &config.model)?;
+
+    match app.parameters {
+        Some(ref parameters) => TaggerTrainer::load_weights(graph, parameters),
+        None => TaggerTrainer::random_weights(graph),
+    }
+}
+
+fn train_model_with_encoder<E>(
+    config: &Config,
+    app: &PretrainApp,
+    trainer: TaggerTrainer,
+    encoder: E,
+    mut train_file: File,
+    mut validation_file: File,
+) -> Result<(), Error>
+where
+    E: SentenceEncoder,
+    E::Encoding: Clone + Eq + Hash,
+    Numberer<E::Encoding>: CborRead,
+{
+    let labels: Numberer<E::Encoding> = config.labeler.load_labels().or_exit(
+        format!(
+            "Cannot load or create label file '{}'",
+            config.labeler.labels
+        ),
+        1,
+    );
+
+    let mut categorical_encoder = CategoricalEncoder::new(encoder, labels);
+
+    let embeddings = config
+        .embeddings
+        .load_embeddings()
+        .or_exit("Cannot load embeddings", 1);
+    let vectorizer = SentVectorizer::new(embeddings);
+
+    let mut best_epoch = 0;
+    let mut best_acc = 0.0;
+
+    let mut save_scheduler = app.save_schedule.into();
+
+    let train_size = train_file.metadata()?.len() as usize;
+
+    for epoch in 0..app.epochs {
+        let (loss, acc) = run_epoch(
+            config,
+            app,
+            &mut save_scheduler,
+            &mut categorical_encoder,
+            &vectorizer,
+            &trainer,
+            &mut train_file,
+            true,
+            train_size,
+            epoch,
+        )?;
+
+        eprintln!("Epoch {} (train, loss: {:.4}, acc: {:.4}", epoch, loss, acc);
+
+        save_scheduler.save(&trainer, CompletedUnit::Epoch).or_exit(
+            format!("Cannot save model for epoch {}", save_scheduler.epoch()),
+            1,
+        );
+
+        let (loss, acc) = run_epoch(
+            config,
+            app,
+            &mut save_scheduler,
+            &mut categorical_encoder,
+            &vectorizer,
+            &trainer,
+            &mut validation_file,
+            false,
+            train_size,
+            epoch,
+        )?;
+
+        if acc > best_acc {
+            best_epoch = epoch;
+            best_acc = acc;
+        }
+
+        let epoch_status = if best_epoch == epoch { "🎉" } else { "" };
+
+        eprintln!(
+            "Epoch {} (validation): loss: {:.4}, acc: {:.4}, best epoch: {}, best acc: {:.4} {}",
+            epoch, loss, acc, best_epoch, best_acc, epoch_status
+        );
+    }
+
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
```

Comment:
Large lists of arguments are something that we see quite often in our tools. Maybe defining a struct that holds the relevant parameters could simplify some of the binaries?

Reply:
I could do that in a separate PR, though I want to investigate first how ad-hoc the structs would be.

Reply:
Wrapping all the parameters in a struct and passing that wouldn't really alleviate anything, maybe it's possible to implement `run_epoch` on a struct. But that's nothing for this PR, i just saw the line while going through the changed files.

Reply:
Yeah, I already felt bad when I disabled that clippy lint ;).